lib/Support/CompilationLog.cpp

```diff
@@ -0,0 +1,58 @@
+/**
+ * Copyright (c) 2017-present, Facebook, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "glow/Support/CompilationLog.h"
+#include "llvm/Support/raw_ostream.h"
+#include "llvm/Support/FileSystem.h"
+
+#include <fstream>
+
+namespace glow {
+llvm::cl::opt<bool>
+    dumpCompilationLogOpt("dump-compilation-log",
+                     llvm::cl::desc("Dump compilation log"));
+
+std::vector<std::string> CompilationScope::scopeStack = {};
+
+CompilationScope::CompilationScope(std::string scopeName) {
+    this->scopeName = scopeName;
+    scopeStack.push_back(scopeName);
+}
+
+CompilationScope::~CompilationScope() {
+    scopeStack.pop_back();
+}
+
+llvm::StringRef CompilationScope::getCurrentCompilationScopeStack() {
+    std::string scopes = "";
+    for (auto scopeStr : scopeStack) {
+        scopes = scopes + "-->" + scopeStr;
+    }
+    return scopes;
+}
+
+void dumpCompilationLog (std::string logContent) {
+    dumpCompilationLog ("compile.log", logContent);
+}
+void dumpCompilationLog (llvm::StringRef logFilename, std::string logContent) {
```

Comment:
Consider using a reference to a string here to avoid copying the log content (which could be quite long).

Reply:
seems that ofstream cannot work with llvm::stringRef

Reply:
[What about `llvm::raw_fd_ostream`?](https://llvm.org/doxygen/classllvm_1_1raw__ostream.html#a2d4592b138e1cf14fa0b8913d0747ac9)

Reply:
works now! Thx a lot!